scripts/change-log-generator.js

```diff
@@ -0,0 +1,248 @@
+const FILES = 'FILES';
+const PACKAGES = 'PACKAGES';
+
+// Regex
+const RELEASE_REGEX = new RegExp(/^\d{2}\.\d{1,2}\.\d/);
+const PR_REGEX = new RegExp(/(\(#\d+\))/);
+const COMMIT_REGEX = new RegExp(/^([\da-zA-Z]+)/);
+
+/**
+ * Gets the latest release branch using git. Retrieves all remote branches
+ * with the name 'origin/release/v*', sorts them in reverse order based on
+ * creation date, and truncates this result so that only the latest branch
+ * is returned.
+ */
+function getLatestReleaseBranch() {
+  var latestReleaseBranch = shell
+    .exec(
+      "git branch --remotes --list --sort='-creatordate' 'origin/release/v*'",
+      { silent: true }
+    )
+    .stdout.trim()
+    .replace('origin/release/v', '')
+    .split(' ', 1)
+    .toString()
+    .replace('\n', '');
+  process.stdout.write('Latest Release Branch: ' + latestReleaseBranch + '\n');
+  return latestReleaseBranch;
+}
+
+function getReleaseBranch() {
+  var releaseIndex = process.argv.indexOf('-r');
+  if (releaseIndex > -1) {
+    return process.argv[releaseIndex + 1];
+  }
+  return getLatestReleaseBranch();
+}
+
+function validateReleaseBranch(releaseBranch) {
+  if (!(releaseBranch && RELEASE_REGEX.exec(releaseBranch))) {
+    process.exit(-1);
+    console.log(
+      "Invalid release '" + releaseBranch + "'. Expected format [xx.yy.z]."
+    );
+  }
+  process.stdout.write('Using release branch: ' + releaseBranch + '\n\n');
+}
+
+function getNewChangeLogBranch(releaseBranch) {
+  var changeLogBranch = CHANGE_LOG_BRANCH + releaseBranch;
+  shell.exec(
+    'git checkout -b ' + changeLogBranch + ' origin/release/v' + releaseBranch
+  );
+}
+
+function getChangeLogText(releaseBranch) {
+  var changeLogText = util.format(LOG_HEADER, releaseBranch.toString());
+  var messagesByPackage = getMessagesPerPackage();
+  Object.keys(messagesByPackage).forEach(function(packageName) {
+    changeLogText += util.format(SECTION_HEADER, packageName);
+    messagesByPackage[packageName].forEach(function(message) {
+      changeLogText += message;
+    });
+  });
+  return changeLogText + '\n';
+}
+
+/**
+ * Groups all messages per package header so they can be displayed under
+ * the same package header subsection. Returns a map of lists.
+ */
+function getMessagesPerPackage() {
+  var messagesByPackage = {};
+  parseCommits().forEach(function(commitMap) {
+    var pr = commitMap[PR_NUM].replace(/[^\d]/g, '');
+    commitMap[PACKAGES].forEach(function(packageName) {
+      messagesByPackage[packageName] = messagesByPackage[packageName] || [];
+      messagesByPackage[packageName].push(
+        util.format(MESSAGE_FORMAT, commitMap[MESSAGE], pr, pr)
+      );
+    });
+  });
+  return messagesByPackage;
+}
+
+/**
+ * Review the log and retrieve all of the commits specific to this
+ * release. Parse this info and return it as a list of hashmaps.
+ */
+function parseCommits() {
+  var results = [];
+  var commits = shell
+    .exec('git log --oneline', { silent: true })
```

Comment:
I found that this approach isn't guaranteed to bring in all of the commits that might be part of a release. For example, if we are committing changes to develop between Tuesday-Thursday, the log of the new release branch shows that these commits appear before the SHA256 that we're checking for. 

This is the approach that I think will work better:
1) git log --left-right --graph --cherry-pick --oneline new_release_branch...previous_release_branch
2) For each of the commits listed, check if their PR# is already a part of the changelog. If they are not, then include them.

Reply:
Note that I would include these changes as part of a separate PR.